### PR TITLE
Feature/db test poc

### PR DESCRIPTION
**Pull request by stmuk**
Add in "comptest" - pkg/geodata/rowquery_test.go is a real test for rowQuery (simple case) - comptests/tests/main_test.go are "cleaner" example tests.

These tests use an empty db (just schema) and put in data so that tests pass.

Refactor schema creation into model.provision.go and add funcs to setup postgres docker and tear down.

### CODE REVIEW

File: comptests/tests/main_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	model.SetupDB(dsn)
+	var err error
+	db, err = gorm.Open(postgres.Open(dsn), &gorm.Config{})
+	if err != nil {
+		log.Print(err)
+	}
+
+}
+
+func TestGorm(t *testing.T) {
```
Comment (VivianAllen) [inline comptests/tests/main_test.go]: is this a test that gorm respects the inner transaction rollback stuff we're using to insulate tests?

File: comptests/tests/main_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	geot := model.GeoType{}
+
+	db.First(&geot)
+
+	if geot.Name == "EW" {
+		t.Fail()
+	}
+}
+
+func TestPgx(t *testing.T) {
```
Comment (VivianAllen) [inline comptests/tests/main_test.go]: likewise, is this a test that PGX allows rollback?

File: model/provision.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	_, pw, host, port, db := ParseDSN(dsn)
+
+	{
+		db, err := gorm.Open(postgres.Open(CreatDSN("postgres", pw, host, port, "postgres")), &gorm.Config{})
+		if err != nil {
+			log.Print(err)
+		}
+
+		// should replace creatdbuser.sh
+		execSQL(db, []string{
```
Comment (VivianAllen) [inline model/provision.go:24]: Would we be able to use this script to setup a new AWS RDS instance?  Guessing this bit with the hardcoded user / password might need tweaking

File: pkg/geodata/rowquery_test.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	"testing"
+
+	"github.com/ONSdigital/dp-find-insights-poc-api/comptests"
+	"github.com/ONSdigital/dp-find-insights-poc-api/model"
+	"github.com/ONSdigital/dp-find-insights-poc-api/pkg/database"
+)
+
+const dsn = "postgres://insights:insights@localhost:54323/censustest"
+
+// passing -args -kill=false to the test will kill docker postgres
```
Comment (VivianAllen) [inline pkg/geodata/rowquery_test.go]: should this say `-kill=true` will kill docker postgres?

File: pkg/geodata/rowquery_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		log.Fatal(err)
+	}
+
+	func() {
+		db.DB().Exec("BEGIN")
+		defer db.DB().Exec("ROLLBACK")
+
+		// setup database
+		doSQL(t, db, "INSERT INTO GEO_TYPE (id,name) VALUES (6,'LSOA')")
+		doSQL(t, db, "INSERT INTO GEO (id,type_id,code,name,lat,long,valid,wkb_geometry,wkb_long_lat_geom) VALUES (7562,6,'E01000001','City of London 001A',51.5181,-0.09706,true,'0103000020E6100000010000000600000099B7D26A2C41B8BFCABAA1E4A2C249403E6F5814C06FB8BF4B7A3CFEF9C149407A98D8707087B9BF11B32F7F25C249403FBA9D552F37B9BF7CB1250DA1C24940ECB58D3C65E6B8BF8C2B844AC3C2494099B7D26A2C41B8BFCABAA1E4A2C24940','0101000020E610000045F0BF95ECD8B8BF5F07CE1951C24940')")
```
Comment (VivianAllen) [inline pkg/geodata/rowquery_test.go]: v. long line here - not sure if thats a fixable issue or not

File: comptests/tests/main_test.go (hunk 1 above)
Comment (stmuk) [reply comptests/tests/main_test.go]: It's an example of how to use gorm transactions.

**Review by VivianAllen — CHANGES_REQUESTED** [review]
- looks like the comment on rowquery_test line 18 might need correcting
- long line on rowquery_test line 37
- could maybe use some explanatory notes on TestGorm and TestPGX in main_test? I wasn't sure what was being tested.
Overall very nice though, looking forward to using it

File: comptests/tests/main_test.go (hunk 2 above)
Comment (stmuk) [reply comptests/tests/main_test.go]: Likewise it's It's an example of how to use pgx transactions. I'd assumed that we were using pgx directly but it's not that simple since the transactions are actually sql.Tx transactions (at least for rowQuery) so I ended up having to use raw transactions on the db handle rather than pgx.

File: model/provision.go (hunk 3 above)
Comment (stmuk) [reply model/provision.go:24]: There is an existing program which does that https://github.com/ONSdigital/dp-find-insights-poc-api/blob/develop/cmd/creatschema/main.go 

Some of that code is refactored into this package. Yes the password shouldn't be hardcoded (it's a dev one) and it needs further work to provision on prod. I'll add a comment to say that.

File: pkg/geodata/rowquery_test.go (hunk 4 above)
Comment (stmuk) [reply pkg/geodata/rowquery_test.go]: The syntax is a bit odd. I'll add a makefile target which kills docker after a run.

File: pkg/geodata/rowquery_test.go (hunk 5 above)
Comment (stmuk) [reply pkg/geodata/rowquery_test.go]: gofmt thought it ok but I'll tweak it.

File: comptests/tests/main_test.go (hunk 1 above)
Comment (VivianAllen) [reply comptests/tests/main_test.go]: ahh gotcha - maybe TestGormExample or something would be clearer? Also, what does it mean if either of these tests fail?

File: comptests/tests/main_test.go (hunk 2 above)
Comment (VivianAllen) [reply comptests/tests/main_test.go]: see comment above for the gorm example - worth thinking about having 'Example' in the name, maybe?

File: comptests/tests/main_test.go (hunk 1 above)
Comment (stmuk) [reply comptests/tests/main_test.go]: see comments below

File: comptests/tests/main_test.go (hunk 2 above)
Comment (stmuk) [reply comptests/tests/main_test.go]: I've added a comment and put Example in the test names.

If these tests fail either docker or postgres isn't running or the geo_type table hasn't been provisioned.

File: pkg/geodata/rowquery_test.go (hunk 4 above)
Comment (stmuk) [reply pkg/geodata/rowquery_test.go]: test-comptest-kill target

File: pkg/geodata/rowquery_test.go (hunk 5 above)
Comment (stmuk) [reply pkg/geodata/rowquery_test.go]: we don't need the geography types in this test set to null

**Review by VivianAllen — APPROVED** [review]
lgtm ✅
